File: crafting/python/src/lox.py
import sys
from parser import Parser

from error_handler import ErrorHandler
from interpreter import Interpreter
from resolver import Resolver
from scanner import Scanner


USAGE_ERROR = 64
BASE_ERROR = 65
RUNTIME_ERROR = 70


class Lox:
    def __init__(self, interpreter: Interpreter) -> None:
        self.interpreter = interpreter

    def run_file(self, path: str) -> None:
        f = open(path)
        self.run(f.read())
        if ErrorHandler.has_error:
            sys.exit(BASE_ERROR)
        if ErrorHandler.had_runtime_error:
            sys.exit(RUNTIME_ERROR)

    def run_prompt(self) -> None:
        while True:
            line = input("rj-plox > ")
            if line is None or line == 'exit':
                break
            self.run(line)
            ErrorHandler.has_error = False

    def run(self, source: str) -> None:
        scanner = Scanner(source)
        tokens = scanner.scan_tokens()
        parser = Parser(tokens)
        statements = parser.parse()
        if ErrorHandler.has_error:
            sys.exit(BASE_ERROR)
        resolver = Resolver(interpreter)
        resolver.resolve_stmts(statements)
        if ErrorHandler.had_runtime_error:
            sys.exit(RUNTIME_ERROR)
        self.interpreter.interpret(statements)


if __name__ == '__main__':
    interpreter = Interpreter()
    lox = Lox(interpreter)
    if len(sys.argv) > 2:
        print('Usage: rjpylox [script name]')
        sys.exit(USAGE_ERROR)
    elif len(sys.argv) == 2:
        script = sys.argv[1]
        lox.run_file(script)
    else:
        lox.run_prompt()

File: crafting/julia/src/Lox.jl
module Lox


export runfile, Runner, runprompt, runstring


mutable struct Runner
    io::IO
    ioerr::IO
    haderror::Bool
    hadruntimeerror::Bool
end


Runner() = Runner(stdout, stderr, false, false)
Runner(io::IO, ioerr::IO) = Runner(io, ioerr, false, false)


include("Errors.jl")
using .Errors

include("Expressions.jl")
include("Interpreters.jl")
using .Interpreters

include("Parser.jl")
using .Parsers

include("Resolvers.jl")
using .Resolvers

include("Scanner.jl")
using .Scanners

include("Statements.jl")
include("Tokens.jl")
using .Tokens


const BASE_ERR = 65
const RUNTIME_ERR = 70


function runfile(runner::Runner, p::AbstractString)
    contents = read(p, String)
    terp = Interpreter(runner)
    run(runner, terp, contents)
    runner.haderror && return BASE_ERR
    runner.hadruntimeerror && return RUNTIME_ERR
    0
end


function runstring(runner::Runner, contents::AbstractString)
    terp = Inerpreter(runner)
    run(runner, terp, contents)
    runner.haderror && return BASE_ERR
    runner.hadruntimeerror && return RUNTIME_ERR
    0
end


function runprompt(runner::Runner)
    terp = Interpreter(runner)
    while true
        print("lox > ")
        line = readline()
        run(runner, terp, line)
        println()
        runner.haderror = false
    end
end


function run(runner::Runner, terp, source::AbstractString)
    scanner = Scanner(runner, source)
    tokens = scantokens(scanner)
    parser = Parser(runner, tokens)
    statements = parser()
    runner.haderror && return
    resolver = Resolver(terp)
    resolve(resolver, statements)
    runner.haderror && return
    interpret(terp, statements)
end


end  # module
